### ISSUE TITLE
some expressions are skipped during mutation phase

### ISSUE THREAD

**Issue by dupdob** [stryker-mutator/stryker-net#351]
Current version (not released) fails to generate mutants for some expressions. Example:
```csharp
namespace Test 
{
    using System;
    public static class CharCheckExtensions
    {
        public static string ChainCalls(string check)
        {
            return check.Replace("ab", "cd")
                .Replace("12", "34")
                .PadLeft(12)
                .Replace("12", "34");
        }
    }
}
```
After reviewing [MutantOrchestrator.cs](https://github.com/stryker-mutator/stryker-net/blob/4a1422a4a9961db562450462316c12b57c8b63fc/src/Stryker.Core/Stryker.Core/Mutants/MutantOrchestrator.cs#L210) I understand that it is due to neglecting the 'Expression' properties for _InvocationExpressionSyntax_. As it represents the target for the invocation, one would logically assume it simply provides some variable name. It turns out that chained calls, as in the previous example, the 'Expression' member can contains quite some code as one can see here:[sharplab](https://sharplab.io/#v2:C4LghgzsA+B2YFsCmEAOYDGSAEAVFw2AsAFADep2V2AAgAy0CMA3JdTQMxMBstATNgDCACzAAnEUgwBrAKIAPYElgQAlgHsVbKhRLV9tLjUa9jDEWFWxBYADa2IACjPYMwqdICU2g7oP/aAHZXdxkAOgAlJFRbTCRHACIwACMEgBpsBIwAEwTvPQDCyOjYrETGPnTMjgAWPJ9C/TCABTBsgBkkADNgRwr8xoDimLjyyoyE2rzWAv8AXx8FkjmgA=).

**Comment by dupdob** [comment]
I am working on a fix.

**Comment by dupdob** [comment]
I am afraid I was optimistic. This issue is harder to fix than anticipate.
I am having a hard time understanding how the Mutate method is supposed to work since it has been refactored.
I also discovered my unit test was wrong. It turns out that above code does mutate:
```csharp
namespace Test 
{
    using System;
    public static class CharCheckExtensions
    {
        public static string ChainCalls(string check)
        {
            return Stryker.ActiveMutationHelper.ActiveMutation==5?check.Replace("ab", "cd")
                    .Replace("12", "34")
                    .PadLeft(12)
                    .Replace("12", "")
:Stryker.ActiveMutationHelper.ActiveMutation==4?check.Replace("ab", "cd")
                    .Replace("12", "34")
                    .PadLeft(12)
                    .Replace("", "34")
:Stryker.ActiveMutationHelper.ActiveMutation==3?check.Replace("ab", "cd")
                    .Replace("12", "")
                    .PadLeft(12)
                    .Replace("12", "34")
:Stryker.ActiveMutationHelper.ActiveMutation==2?check.Replace("ab", "cd")
                    .Replace("", "34")
                    .PadLeft(12)
                    .Replace("12", "34")
:Stryker.ActiveMutationHelper.ActiveMutation==1?check.Replace("ab", "")
                    .Replace("12", "34")
                    .PadLeft(12)
                    .Replace("12", "34")
:Stryker.ActiveMutationHelper.ActiveMutation==0?check.Replace("", "cd")
                    .Replace("12", "34")
                    .PadLeft(12)
                    .Replace("12", "34")
:check.Replace("ab", "cd")
                    .Replace("12", "34")
                    .PadLeft(12)
                    .Replace("12", "34")
                ;
        }
    }
}
```
which is needlessly complex, but generates expected mutants. I will go back to the analysis and devise a test that actually reproduce the issue.

Meanwhile: I may be wrong, but I think that those mutants would benefit being written this way:
```csharp
namespace Test 
{
    using System;
    public static class CharCheckExtensions
    {
        public static string ChainCalls(string check)
        {
            return check.Replace(
Stryker.ActiveMutationHelper.ActiveMutation==0? "" :"ab", 
Stryker.ActiveMutationHelper.ActiveMutation==1?  "" : "cd")
                    .Replace(
Stryker.ActiveMutationHelper.ActiveMutation==2? " : "12", 
Stryker.ActiveMutationHelper.ActiveMutation==3? "": "34")
                    .PadLeft(12)
                    .Replace(
Stryker.ActiveMutationHelper.ActiveMutation==4? "" : "12",  Stryker.ActiveMutationHelper.ActiveMutation==5? "" : "34")
                ;
        }
    }
}
```
Does it make sense and is it realistic? I will open an issue if you think it is worth it

**Comment by richardwerkman** [comment]
I tried to mutate the arguments directly but this looks harder than it seems... For now the current way of mutating the return value seems fine by me. @dupdob are you still looking into why there are mutations missing.

**Comment by dupdob** [comment]
I am, but so far, I mostly progressed in my understanding of the logic, not much progress toward resolution. Based on your comment, I have the feeling that the MutationOrchestrator. Mutate has a somewhat chaotic behaviour, as I get mutated arguments or mutated expressions on two very similar bits of code:

```csharp
namespace Test 
{
    using System;
    public static class CharCheckExtensions
    {
        public static ICheckLink<ICheck<char>> IsALetter(this ICheck<char> check)
        {
            ExtensibilityHelper.BeginCheck(check).
                FailWhen(sut => !IsALetter(sut), "The {0} is not a letter.").
                OnNegate("The {0} is a letter whereas it must not.").
                EndCheck();
            return ExtensibilityHelper.BuildCheckLink(check);
        }
        public static string ChainCalls(string check)
        {
            return check.Replace("ab", "cd")
                .Replace("12", "34")
                .PadLeft(12)
                .Replace("12", "34")
                .Normalize();
        }
    }
}
```
Mutates to :
```csharp
namespace Test 
{
    using System;
    public static class CharCheckExtensions
    {
        public static ICheckLink<ICheck<char>> IsALetter(this ICheck<char> check)
        {
            ExtensibilityHelper.BeginCheck(check).
                FailWhen(sut => !IsALetter(sut), "The {0} is not a letter.").
                OnNegate(Stryker.ActiveMutationHelper.ActiveMutation==0?"":"The {0} is a letter whereas it must not.").
                EndCheck();
            return ExtensibilityHelper.BuildCheckLink(check);
        }

        public static string ChainCalls(string check)
        {
            return Stryker.ActiveMutationHelper.ActiveMutation==6?check.Replace("ab", "cd")
                .Replace("12", "34")
                .PadLeft(12)
                .Replace("12", "")
                .Normalize():Stryker.ActiveMutationHelper.ActiveMutation==5?check.Replace("ab", "cd")
                .Replace("12", "34")
                .PadLeft(12)
                .Replace("", "34")
                .Normalize():Stryker.ActiveMutationHelper.ActiveMutation==4?check.Replace("ab", "cd")
                .Replace("12", "")
                .PadLeft(12)
                .Replace("12", "34")
                .Normalize():Stryker.ActiveMutationHelper.ActiveMutation==3?check.Replace("ab", "cd")
                .Replace("", "34")
                .PadLeft(12)
                .Replace("12", "34")
                .Normalize():Stryker.ActiveMutationHelper.ActiveMutation==2?check.Replace("ab", "")
                .Replace("12", "34")
                .PadLeft(12)
                .Replace("12", "34")
                .Normalize():Stryker.ActiveMutationHelper.ActiveMutation==1?check.Replace("", "cd")
                .Replace("12", "34")
                .PadLeft(12)
                .Replace("12", "34")
                .Normalize():check.Replace("ab", "cd")
                .Replace("12", "34")
                .PadLeft(12)
                .Replace("12", "34")
                .Normalize();
        }
    }
}
```

Why such a variation ?? probably because the first one uses a static method will the second one use an instance method....

**Comment by richardwerkman** [comment]
This has nothing to do with static or instance methods. It has to do with the kind of syntaxnode wraps the stringliteral. 

In `ICheckLink` the stringliteral is present in a `SimpleMemberAccessExpression` and in `ChainCalls` the strings are present in a `ReturnStatement`. Both `ReturnStatement` and `SimpleMemberAccessExpression` contain expressions but one does call mutate recurively and one mutates the whole expression with conditionalexpressions.

The mutantorchestrator is getting kind of messy I agree. But there are a lot of cases to keep track of and they are all a bit different.

**Comment by dupdob** [comment]
thanks fo the pointer 👍 . I understand where the complexity of Mutate is coming from, I was venting some frustration from my inability to get a decent grasp on its logic.

**Comment by dupdob** [comment]
Still working on this. I think I go my head around the syntax tree parsing. 
I still need to improve my grasp on Roslyn immutability and what it implies
Slowly getting there

**Comment by richardwerkman** [comment]
The immutablilty of the syntax trees was a big challenge. If it helps I can try to create a decision tree of the logic. It might help to understand what's going on. If you find a certain node kind that is not mutated anymore please state that in this issue. If we have a reproduction case we can look more specific to why there are less mutations being made.

**Comment by dupdob** [comment]
Thanks for the offer. Honestly, I wanted to contribute to Stryker partly because it gave me an opportunity to look into roslyn, so perfect training for me (the other parts being because I need it, and I believe in mutation testing)
In terms of parsing logic, it looks like there are issues for complex expressions where only one subexpression is mutated. It depends on where the expression is located.

So I am attempting to redesign **Mutate** to handle interesting statements first, then expressions (incl. subexpressions) and simple recursion when everything else failed

**Comment by dupdob** [comment]
Morning update: I made significant progress.
current release: 1717 mutants, 1604 ready (4 failed rollbacks)
head of repo: 933 mutants, 898 ready (2 failed rollbacks)
my WIP: 1693 mutants, 1566 ready (0 failed rollback)
so still missing: ~25 initial mutants, and ~40 ready.

Next steps: clean up/refactor code I produced then look into mutants that could no compile

**Comment by richardwerkman** [comment]
Sounds great! A few commits ago #341 we removed the "remove linq statement" mutations. This could explain some missing mutations. I think that once we release the html reporter we can see more clearly what pieces of code we are not mutating yet. With the current console report the missing parts are harder to spot. I'm looking forward to your PR

**Comment by dupdob** [comment]
Morning update, current stats:
current release: 1717 mutants, 1604 ready (4 failed rollbacks)
head of repo: 933 mutants, 898 ready (2 failed rollbacks)
**my WIP: 1656 mutants, 1565 ready (0 failed rollback)**

I am currently focused on mutants failing to build  because:
1. Some can be fixed as they fails due to bad control structure (conditional operator instead of if)
2. The other ones add noise to the build and can lead to failed rollback due to error hiding

Regarding (1), I have an issue wit **for** statements, as my current mutation algorithm uses conditional construct as in
```csharp
// original
for( var i =0; i<10; i++) {}
// mutant
for (var i=0; ... ? i<10 : i<=10; ... ? i++ : i--) {}
```
Which does not compile for obvious reason.

Regarding (2): I added filtering logic for string concatenation, so I no longer mutate
```csharp
var demo = "some" + "text";
// no longer mutated to
var demo = "some" - "text";
```
Filtering is made on + when any argument is a string , and on += when right argument is a string (string here meaning either a literal string or an interpolated string).

Random remark:
I have lost one (1) valid mutant (1566 -> 1565) for no obvious reason.

**Comment by dupdob** [comment]
Morning update: getting there!
current release: 1717 mutants, 1604 ready (4 failed rollbacks)
head of repo: 933 mutants, 898 ready (2 failed rollbacks)
**my WIP: 1615 mutants, 1564 ready (0 failed rollback)**

- I fixed the problem with for statement I discovered yesterday.
- lost another mutant in the process (1365=> 1364), which is unexpected as I must have restored some CompileError mutants.
- Added some checks to reduce the number of **CompileError**s
- I ran into #353 : it looks like most, if not all mutants, no longer have valid line numbers. Will deal with it later
- made a side by side analysis of run reports and spotted missing checks

I will add unit tests on relevant code block to understand why the analysis skips part of the expression.

**Comment by richardwerkman** [comment]
Do you have some code available for us to look at? We are curious to what changes you made and what the effects are!

**Comment by dupdob** [comment]
Morning update: **Looking good**
current release: 1717 mutants, 1604 ready (4 failed rollbacks)
head of repo: 933 mutants, 898 ready (2 failed rollbacks)
**my WIP: 1723 mutants, 1664 ready (0 failed rollback)**
I am happy to report I have been able to increase the total number of generated mutants, in regard to the previous release.

I DO NOT CONSIDER WORK AS DONE YET!
I will open a PR for you to review work in progress.

**Comment by richardwerkman** [comment]
Hi, I've looked at your PR and indeed it's looking great! Mutating seems smooth and without errors. I'll quickly give some feedback on the PR. Maybe it can help :)

**Comment by dupdob** [comment]
I have found the issue with the integration tests. It should be ok now.
But I found another issue while testing on NFluent... keep on working

**Comment by dupdob** [comment]
Ok. I fixed the issue I had with for statements.
Latest numbers are good,

**Comment by dupdob** [comment]
__Note: Previous close was a mishap__
Morning update: **almost there**
current stryker release: 1717 mutants, 1604 ready (4 failed rollbacks)
head of repo: 933 mutants, 898 ready (2 failed rollbacks)
### my WIP: 1727 mutants, 1707 ready (0 failed rollback) 
I am really happy with the current status: got 100+ new mutants, no failed rollback, only a handful of CompileError (remaining causes are : undetectable string concatenation and LinqMutants on on Linq methods).
Location data has been restored for most mutants, some remaining to be done.

**Comment by dupdob** [comment]
Morning update: 
current stryker release: 1717 mutants, 1604 ready (4 failed rollbacks)
head of repo: 933 mutants, 898 ready (2 failed rollbacks)
## WIP: 1763 mutants, 1744 ready (0 failed rollback)
 
Further refactoring and simplification provided some unexpected boost.
So total progress (for NFluent) is around +10%

**Comment by Mobrockers** [comment]
@dupdob Do you believe the linq mutator is actually useful now with the new changes? We had recently removed it because we deemed it to mostly generate false positives and cause rollbacks.

**Comment by dupdob** [comment]
No opinion. I kinda like having a Linq mutator. I simply added a check to make sure the LinqMutator mutated a method name and skipped attributes/properties, thereby removing a bunch of CompileError.

So my change increased the efficiency of this mutator.

I only started recently to use Linq, but I know that many C# dev use it intensively.
I would be interested in discussing false positive.

Considering what I have written, I would recommend keeping it but looking into ways to improve its efficiency.
